iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergBucketFunction.java

```diff
@@ -0,0 +1,169 @@
+    Objects.requireNonNull(numBuckets);
+    Preconditions.checkArgument(sourceColumnNames.size() == numBuckets.size());
+    Preconditions.checkArgument(!sourceColumnNames.isEmpty());
+    this.sourceColumnNames = sourceColumnNames;
+    this.numBuckets = numBuckets;
+  }
+
+  @Override
+  public Optional<CustomBucketFunction> select(boolean[] retainedColumns) {
+    Preconditions.checkArgument(retainedColumns.length == numBuckets.size());
+    Preconditions.checkState(bucketUdfs == null);
+    Preconditions.checkState(bucketIdInspectors == null);
+    Preconditions.checkState(deferredObjects == null);
+
+    final List<String> newSourceColumnNames = Lists.newArrayList();
+    final List<Integer> newNumBuckets = Lists.newArrayList();
+    for (int i = 0; i < retainedColumns.length; i++) {
+      if (retainedColumns[i]) {
+        newSourceColumnNames.add(sourceColumnNames.get(i));
+        newNumBuckets.add(numBuckets.get(i));
+      }
+    }
+
+    if (newSourceColumnNames.isEmpty()) {
+      return Optional.empty();
+    }
+
+    final IcebergBucketFunction newBucketFunction = new IcebergBucketFunction(newSourceColumnNames, newNumBuckets);
+    return Optional.of(newBucketFunction);
+  }
+
+  @Override
+  public void initialize(ObjectInspector[] objectInspectors) {
+    Preconditions.checkArgument(objectInspectors.length == numBuckets.size());
+    bucketIdInspectors = Lists.newArrayList();
+    bucketUdfs = Lists.newArrayList();
+    for (int i = 0; i < objectInspectors.length; i++) {
+      final GenericUDFIcebergBucket udf = new GenericUDFIcebergBucket();
+      final ObjectInspector numBucketInspector = PrimitiveObjectInspectorFactory
+          .getPrimitiveWritableConstantObjectInspector(
+              TypeInfoFactory.intTypeInfo,
+              new IntWritable(numBuckets.get(i)));
+      final ObjectInspector[] args = { objectInspectors[i], numBucketInspector };
+      try {
+        final ObjectInspector inspector = udf.initialize(args);
+        Preconditions.checkState(inspector.getCategory() == Category.PRIMITIVE);
+        final PrimitiveObjectInspector primitiveInspector = (PrimitiveObjectInspector) inspector;
+        Preconditions.checkState(primitiveInspector.getPrimitiveCategory() == PrimitiveCategory.INT);
+        bucketIdInspectors.add((IntObjectInspector) primitiveInspector);
+      } catch (UDFArgumentException e) {
+        throw new IllegalArgumentException("The given object inspector is illegal", e);
+      }
+      bucketUdfs.add(udf);
+    }
+    deferredObjects = new SettableDeferredObject[1];
+    deferredObjects[0] = new SettableDeferredObject();
+
+    Preconditions.checkState(bucketIdInspectors.size() == numBuckets.size());
+    Preconditions.checkState(bucketUdfs.size() == numBuckets.size());
+  }
+
+  @Override
+  public List<String> getSourceColumnNames() {
+    return sourceColumnNames;
+  }
+
+  @Override
+  public int getNumBuckets() {
+    return numBuckets.stream().reduce(1, (a, b) -> a * b);
```

Comment:
`reduce(1, Math::multiplyExact)`

Reply:
Good to know the method. Done
https://github.com/apache/hive/pull/5409/commits/41ea3ab3b78b2fc88e30a4fcd641540188819e60